### PR TITLE
New page: Add a Debit Card

### PR DESCRIPTION
**Pull request by rdjuric**
### Fixed Issues

$ https://github.com/Expensify/App/issues/3897

### Tests

1. Open your settings
2. Open the "Payments" page
3. There's a new option "Debit Card" under "Add Payment Method"
4. Clicking on the option opens the "Add Debit Card" page

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/53711423/128876868-dad6ecdb-bba5-404f-89ec-06df74c81296.mov

#### Mobile Web

https://user-images.githubusercontent.com/53711423/128876890-afb4d988-5a3c-483a-8a6f-d5a874f2fb57.mp4

#### Desktop

https://user-images.githubusercontent.com/53711423/128876900-7ee3fd39-256e-453f-8452-6a26b4fa99ee.mov

#### iOS

https://user-images.githubusercontent.com/53711423/128876913-423d7452-db11-4e8c-be1c-f595488de363.mp4

#### Android

https://user-images.githubusercontent.com/53711423/128878913-b90eba9b-7bd2-4385-ae85-411126fd8eb0.mp4

### CODE REVIEW

**Comment by rdjuric** [comment]
ccing @stitesExpensify

File: src/libs/actions/PaymentMethods.js
Comment (bondydaa) [inline src/libs/actions/PaymentMethods.js:42]: Isn't the regex for the year expecting just `MM/YY`? this is grabbing 4 digits 
image

Either way you could probably simplify this to just grab everything from the 3 digit and on with:
```suggestion
        cardYear: params.expirationDate.substr(3),
```

**Review by bondydaa — CHANGES_REQUESTED** [review]
Just a few comments so far

File: src/pages/settings/Payments/PaymentsPage.js
Comment (stitesExpensify) [inline src/pages/settings/Payments/PaymentsPage.js:85]: :smile:

File: src/pages/settings/Payments/AddDebitCardPage.js
Comment (stitesExpensify) [inline src/pages/settings/Payments/AddDebitCardPage.js:52]: This is unused.

File: src/libs/ValidationUtils.js
Comment (stitesExpensify) [inline src/libs/ValidationUtils.js:14]: can this be `validateCardNumber` please?

**Review by stitesExpensify — CHANGES_REQUESTED** [review]

**Comment by rdjuric** [comment]
Updated @stitesExpensify!

File: src/libs/ValidationUtils.js
Comment (stitesExpensify) [reply src/libs/ValidationUtils.js:95]: I don't think we need it. I think we're good

File: src/CONST.js
Comment (stitesExpensify) [reply src/CONST.js:377]: Weird... looks like it works correctly though!

File: src/libs/API.js
Comment (stitesExpensify) [inline src/libs/API.js:336]: This line needs to be `return Network.post(commandName, parameters, CONST.NETWORK.METHOD.POST, true);` because the command is on our secure server, not the normal one.

File: src/libs/ValidationUtils.js
Comment (stitesExpensify) [inline src/libs/ValidationUtils.js:72]: We actually need to send this in the MM/YYYY format. Can we change this to be a function that accepts both MM/YYYY, and MM/YY, but converts MM/YY into MM/YYYY automatically?

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:44]: This needs to specifically be `cardCVV`

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:46]: We need to add a new param here `currency: 'USD'`

File: src/pages/settings/Payments/AddDebitCardPage.js
Comment (stitesExpensify) [inline src/pages/settings/Payments/AddDebitCardPage.js:143]: Can we make it so that when you're typing we automatically put in a `/` after the first 2 digits? Not sure how hard that would be to implement tbh, but I think it's a big quality of life improvement

**Comment by stitesExpensify** [comment]
I just tested locally and there were just a couple of small issues with the API call (which you obviously couldn't have known, that's on me). The only other request is a quality of life improvement that I think is very valuable. Almost at the finish line!

**Comment by stitesExpensify** [comment]
Hi @rdjuric any update here?

**Comment by rdjuric** [comment]
Sorry, got sick. Updating this today!

**Comment by stitesExpensify** [comment]
No problem, glad you're feeling better. Thanks for the update!

File: src/libs/ValidationUtils.js
Comment (rdjuric) [reply src/libs/ValidationUtils.js:72]: I think we can modify this validation to accept both MM/YY and MM/YYYY, but leave the conversion to another function that we'll call in the `addBillingCard` API.

File: src/pages/settings/Payments/AddDebitCardPage.js
Comment (rdjuric) [reply src/pages/settings/Payments/AddDebitCardPage.js:143]: Agreed! Good idea

**Comment by rdjuric** [comment]
PR updated @stitesExpensify ! Sorry again for the delay.

File: src/CONST.js
Comment (bondydaa) [reply src/CONST.js:377]: wow yeah very strange, might have to remove regexper from my toolbelt if it can't be trusted.

Comment (bondydaa) [reply src/CONST.js:377]: image

ooooh how dumb, if you hover over the repeats it shows why - it's including the numbers shown as a "pass" which I guess sort of makes sense but is also super confusing 🤦

File: src/libs/actions/PaymentMethods.js
Comment (bondydaa) [inline src/libs/actions/PaymentMethods.js:73]: I'm not a big fan of this, I don't think we should be modifying user input before sending it to the API.

If we need a 4 digit year then let's just update the regex, placeholder, and input to all expect a 4 digit year and have the user input it themselves. 

Having all this extra code is cumbersome and will be harder to debug than just asking the user to input 2 extra digits.

File: src/pages/settings/Payments/AddDebitCardPage.js
Comment (bondydaa) [inline src/pages/settings/Payments/AddDebitCardPage.js:75]: any reason we're not adding this to the constant file so it could be reused and then creating a method in the utlis?

**Review by bondydaa — DISMISSED** [review]

File: src/libs/actions/PaymentMethods.js
Comment (rdjuric) [reply src/libs/actions/PaymentMethods.js:73]: The first implementation followed the specs (2 digits), but I was asked to change it to 2 or 4 digits (and to convert from 2 to 4). I'll wait for @stitesExpensify to reply before changing this yet again

File: src/pages/settings/Payments/AddDebitCardPage.js
Comment (rdjuric) [reply src/pages/settings/Payments/AddDebitCardPage.js:75]: That's a good idea! Thanks, doing it now.

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [reply src/libs/actions/PaymentMethods.js:73]: I personally like doing both. Just because some cards have the date written on them as `mm/yyyy` and some are just `mm/yy` (I think most are in the latter format). Our API requires a 4 digit year but I don't see a need to pass that requirement onto the user. 

I'm not really worried about the debugging either since this function is so simple. Thoughts @bondydaa?

Comment (bondydaa) [reply src/libs/actions/PaymentMethods.js:73]: I mean this is not easy code to follow IMO and add so much complexity for such little gain - it's not hard for someone to enter 2021 vs 21 when reading it off their card, specially if we're just requiring you do enter 4 digits. Just be we _can_ code something doesn't mean we _should_.

The same result can be achieved by ensuring MM/YYYY with a `pattern="\d{2}\/\d{4}":` on the input which reduces the complexity of the expiration regex in the const file (or allows us to remove it all together). If a user screws up entering their expiration date then we'll return that as a nice error for them to correct and it's on them.

If we have that client side validation then we remove this normalize method all together then which means we're not editing a user's input before it's submitted and have reduced the amount of code we need to maintain in the future.

If you give users wiggle room to fuck something up they will and if we're editing what they input then our logs will show something different than what they gave us and it won't be clear from logs that we edited their entry unless you look at the code. 

Ultimately I'm not going to be the one to maintain any of this code so don't consider this a blocker (it will just be another reason I won't want to touch this code base 😅).

Comment (stitesExpensify) [reply src/libs/actions/PaymentMethods.js:73]: Alright, sorry for the back and forth @rdjuric but let's just require `MM/YYYY` to simplify the code. After that I'm thinking this is good to go!

Comment (rdjuric) [reply src/libs/actions/PaymentMethods.js:73]: No worries! It's part of the process 😄

**Comment by rdjuric** [comment]
PR updated! @stitesExpensify

**Comment by stitesExpensify** [comment]
We need to update this placeholder (i typed it wrong like 5 times and was so confused :joy:)

**Comment by stitesExpensify** [comment]
2 things here: 
1. Can we make the "add card" button show a loading spinner while the API call is going?
2. It looks like we have some weird behavior with our payment methods after adding a card. I think it's a 2 fold problem between onyx, and us not having the new card's info. I need to ask internally about this because really we should be sending the debit card info back with the createFund call but we're only sending a tiny bit of info. The workaround would be to make another API call that gets the full debit card info


https://user-images.githubusercontent.com/42391420/133531012-0f216d9d-9c8c-4ed8-8e56-33bd4328b905.mp4

**Comment by stitesExpensify** [comment]
BTW you can test adding a card using these https://stripe.com/docs/testing

**Comment by rdjuric** [comment]
>We need to update this placeholder (i typed it wrong like 5 times and was so confused 😂)

Sorry, totally forgot about this 😭

>Can we make the "add card" button show a loading spinner while the API call is going?

 Good idea! Implementing this now.

>It looks like we have some weird behavior with our payment methods after adding a card

Is this a blocker here? We're sending the correct request and just saving the response in ONYX. Should just be a question of sending the correct response in the backend.

**Comment by rdjuric** [comment]
PR updated @stitesExpensify

**Comment by stitesExpensify** [comment]
Ahhh that's the problem! Will comment in the code

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:51]: So right here we're overwriting the entire `ONYXKEYS.CARD_LIST` key with the response from the call which is just this:
```
{
fundID: 21
jsonCode: 200
requestID: "2icDJ6"
}
```

What we actually want to do is use the information we were originally given to add a new card to the list.

The format the cards are stored in is:
```
{
  "additionalData": {
    "isBillingCard": true,
    "isP2PDebitCard": false
  },
  "addressName": "Asdf Asdf", //User's name on the card
  "addressState": "US",
  "addressStreet": "",
  "addressZip": 80026,
  "cardMonth": 9,
  "cardNumber": "520082XXXXXX8210",
  "cardYear": 2023,
  "created": "2021-09-16 00:42:18",
  "currency": "USD",
  "fundID": 21, //This is returned from the api as mentioned above
}
```

So when we get a response that is 200, we should do an `Onyx.merge(<card object in the format above>)`

Comment (stitesExpensify) [reply src/libs/actions/PaymentMethods.js:51]: The only info we don't have is the `created` which is fine, we shouldn't need that.

Comment (stitesExpensify) [reply src/libs/actions/PaymentMethods.js:51]: We also need to mask the card number they input to match the format above

**Comment by rdjuric** [comment]
Thanks for the help @stitesExpensify! I'm having a little trouble testing the end point (getting a CORS error), but I implemented the changes you described.

**Comment by stitesExpensify** [comment]
Cool i'll give it a test today

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:56]: I think these are backwards in this case, we don't want it to be their old expensify billing card, and it _is_ a p2p debit card.

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:64]: We need to mask all but the first 4 and last 4 digits here. There might be a util somewhere I'm not sure

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:69]: This needs to be a merge, not set. We want to add this card to the card list, not set it to only this card

**Review by stitesExpensify — CHANGES_REQUESTED** [review]

File: src/libs/actions/PaymentMethods.js
Comment (rdjuric) [reply src/libs/actions/PaymentMethods.js:64]: Do you mean we want to save the card in ONYX in the format 4242 **** **** 4242?

Comment (stitesExpensify) [reply src/libs/actions/PaymentMethods.js:64]: Yes but with x's like the example above. `"520082XXXXXX8210"`

**Comment by rdjuric** [comment]
Updated @stitesExpensify

**Comment by stitesExpensify** [comment]
Code is looking great! I am getting 2 console errors though.
One of them is when I open the "add payment method" dropdown the keys seem to be broken


The other one is when we try to do the merge something isn't working. I think it might be the format it's passed in? I'm not really sure. Either way it's not merging correctly though

**Comment by rdjuric** [comment]
> One of them is when I open the "add payment method" dropdown the keys seem to be broken

Looks like this is coming from the `PaymentMethodList`, not from the dropdown (our dropdown uses a `Popover` not a virtualized list, so it doesn't use/needs keys). 

From the error message we see that the key being passed to the `PaymentMethodList` is `card-undefined`. I think this is happening because we expect the card to have a `cardID` property here
https://github.com/Expensify/App/blob/d92ac1fc5eda539f76ef88874b4d5a4e8dc172ee/src/pages/settings/Payments/PaymentMethodList.js#L118
but we're not adding one when we merge the new card to Onyx. (I don't see this property in the card format that you gave me, so we're not saving it)

> The other one is when we try to do the merge something isn't working. I think it might be the format it's passed in? I'm not really sure. Either way it's not merging correctly though

I think this is the same problem from above - we're merging the wrong object to Onyx. I don't think I have any way of knowing what is the correct format here.

Is the format expected by `ONYXKEYS.CARD_LIST` documented anywhere? If not, I think we can find the format by checking this [endpoint](https://github.com/Expensify/App/blob/d92ac1fc5eda539f76ef88874b4d5a4e8dc172ee/src/libs/actions/PaymentMethods.js#L12) in the backend. Whatever this endpoint returns under `fundList` we directly save to `ONYXKEYS.CARD_LIST`, so should be in the right format 🧐

Let me know if i'm missing something @stitesExpensify

**Comment by stitesExpensify** [comment]
>  I think this is happening because we expect the card to have a cardID property here

Let's just change that key to the masked card number then

> I think this is the same problem from above - we're merging the wrong object to Onyx. I don't think I have any way of knowing what is the correct format here 🥲

You should be able to look at the card object format by refreshing the page and looking at the card object in the application memory (since that info will be from the server). Onyx is all front end so we should be able to figure it out without diving into any server code

**Comment by stitesExpensify** [comment]
Hey @rdjuric! Any update here?

**Comment by rdjuric** [comment]
> Let's just change that key to the masked card number then

Done

> You should be able to look at the card object format by refreshing the page and looking at the card object in the application memory (since that info will be from the server). Onyx is all front end so we should be able to figure it out without diving into any server code

Unfortunately I don't have any cards saved locally. I couldn't find any way to add a card so that it's returned by the `getPaymentMethods` API and I can check its local object. 

It's hard for me to test this flow since our `AddBillingCard` endpoint isn't available and just returns a CORS error.

**Comment by stitesExpensify** [comment]
It looks like this might solve your CORS issue https://expensify.slack.com/archives/C01GTK53T8Q/p1631690289275000

**Comment by stitesExpensify** [comment]
I found the solution!

File: src/libs/actions/PaymentMethods.js
Comment (stitesExpensify) [inline src/libs/actions/PaymentMethods.js:70]:
```suggestion
            Onyx.merge(ONYXKEYS.CARD_LIST, [cardObject]);
```

**Comment by stitesExpensify** [comment]
If you can fix that line, and the conflicts we can merge this!

**Comment by stitesExpensify** [comment]
Code looks good! Testing and then will merge assuming everything works well

**Comment by stitesExpensify** [comment]
🟢  on testing :tada:

**Review by stitesExpensify — APPROVED** [review]

**Comment by rdjuric** [comment]
Thanks @stitesExpensify! Helped me a lot with this one 🎉

**Comment by Jag96** [comment]
@rdjuric it looks like the inputs used here don't follow the same standard as in other parts of the app, can you update them to match the inputs in the VBA flow? More context here: https://github.com/Expensify/App/issues/5644#issuecomment-933976415.

**Pull request merged**
